src/java.base/share/classes/java/io/SequenceInputStream.java

```diff
@@ -79,10 +80,7 @@ public SequenceInputStream(Enumeration<? extends InputStream> e) {
      * @param   s2   the second input stream to read.
      */
     public SequenceInputStream(InputStream s1, InputStream s2) {
```

Comment:
BTW, what is your opinion @jaikiran  and @AlanBateman: We could simplify the 2-arg constructor by calling `this(...)` instead of repeating the 1-arg constructor's implementation here. Is that a *preferred* or a *disliked* pattern in OpenJDK?

Reply:
The updated code now changes the behaviour in the other direction:

Previously, if `s2` was null a NPE was thrown in `peekNextStream` when `s1` was exhausted.

In the current code, `s2` is silently ignored if it is null.

A safer alternative that preserves the behaviour of nulls seems to be the replace `List.of` with `Arrays.asList`.

These subtle changes in behaviour demonstrates the problem with even trivial updates to legacy code...

Reply:
It depends on *how far* we want to align the behavior. I do see a benefit in accepting `s2` being `null`. I do not see a benefit in throwing NPE at a later time. Why should an application want to expect that?

Reply:
@AlanBateman WDYT?

Reply:
> @AlanBateman WDYT?

The long standing and undocumented behavior is unfortunate. I don't think the 1-arg constructor is fixable while still allowing for lazy behavior. So I think the only thing we can do is document the existing behavior that a null element in the enumeration will lead to read or close throwing NPE at a later time. For the 2-arg constructor then we need to be cautious. This is very old API and we have to assume the phenomenon that is Hyrum's Law. You may be right that nobody can reasonably depend on the current behavior but I think it would need a survey to inform the discussion. I can't tell if this PR was intended to rehabilitate SequenceInputStream or just a reaction to seeing that it uses Vector. If the latter then you could just change it to:
```
var list = new ArrayList<InputStream>();
list.add(s1);
list.add(s2);
e = Collections.enumeration(list);
peekNextStream();
```
and preserve the existing behavior.

While SequenceInputStream may be legacy I guess it would be interesting to get some idea of usage or usages that use something better in libraries. Maybe there is a case for InputStream to have a concat method for example.

Reply:
Indeed my intention solely is to get rid of `Vector`, so I am fine with keeping a low profile and being full backwards compatible. I assume SIS is seldomly used, so opening a can of warms is not worth it (I think). I am perfectly happy with your proposal, but for the sake of brevity I'd rather go with a shorter way of doing the pretty same:
```this(Collections.enumeration(Arrays.asList(s1, s2))```
Are we all fine with that? Otherwise I'd switch this PR back to Alan`s proposal.

Reply:
N.B.: Regarding usage numbers, I will do a quick poll on Twitter.

Reply:
> Indeed my intention solely is to get rid of `Vector`, so I am fine with keeping a low profile and being full backwards compatible. I assume SIS is seldomly used, so opening a can of warms is not worth it (I think).

If you have the cycles then collecting data on usages and finding examples where the constructor is called with a second parameter of null would be useful. It's a lot of work to do that of course. The changes in update 2e957354 look okay, I can't immediately see any behavior change with that version.

Reply:
I think the public visibility of my Twitter account is not wide enough to get *really robust* answers, unfortunately. So far, 92% answered that they not even used SIS in their whole life. So the users of two-args constructor must be really neglectable. Nevertheless, they are definitively not zero, so all we could do is marking it deprecated some day.

Anyways, thanks for the review. At least we get rid of `Vector` here. Would be happy if you mark the PR as reviewed. :-)

Reply:
One alternative is to search GitHub. It's amazing how fast they can search such a huge code corpus.

Example: https://github.com/search?l=&q=%22new+SequenceInputStream%22+-filename%3ASequenceInputStreamTest.java&type=code

One problem with GitHub search is that often you get a lot of duplicate matches. In the example above I have filtered out `SequenceInputStreamTest.java` which showed up a lot.

Reply:
Twitter survey results after one week: 5.4% of all voters have actually used SIS before! This number is big enough to consider SIS as "not irrelevant". OTOH this number is so small that any further poll whether this 5.4% actually used the one-arg or two-arg constructor would not provide stable results. Looking at the results of @jensli's GitHub search tells us that apparently *all* of those findings are using 2-args. So we actually need to be careful, not to break (up to) 5.4% of all existing applications.

So what could be our plan to improve the live of that 5.4%? Some thoughts:
1. We could safely add a `public InputStream concat(InputStream)` *default method* to `InputStream` as a factory for *anonymous* sequences. This allows any IS implementation to provide an optimized variant if needed/wanted, falling back to the existing SIS *by default* as a starting point (until we come up with a better fallback implementation).
2. Having done that, we could deprecate-for-removal SIS with the clear note that from now on `concat` should be used instead. This triggers people to overhaul their existing code little by little, while we gain time to work on a replacement for SIS.
3. We then could provide a **non-public** `ConcatenatedInputStream` which holds **exactly two** IS (`first` and `last` MUST be given), and make that one the default implementation of `concat`, gaining a modern, easy-to-maintain, fully tested, high-performance, sequence of streams.
4. Some fine day we do remove SIS, effectively getting rid of the discussed quirky behavior.

@AlanBateman How does that sound? Shall I go on with a PR for step 1.?

Reply:
I did a search of 30M classes in 130k artifacts from Maven Central and it found usages in 1670 classes in 852 artifacts. I haven't dug into it too closely but it looks like the usages are mostly (if not all) the 2-arg constructor.

If you have the cycles, then prototyping a _concat_ method (you could try both instance and static) would help inform on issues and whether this is worth doing. Adding methods to older, and widely used, classes such as InputStream will often encounter compatibility issues. For the instance method you'll need to think about security/snooping issues. I think it would also be useful to survey libraries that provide an equivalent and if there is much usage.

SIS may be legacy but I don't think there is a case to deprecate it at this time.